Sources/GRPCCore/MethodConfigurations.swift

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2023, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/// A collection of ``MethodConfiguration``s, mapped to specific methods or services.
+///
+/// When creating a new instance, no overrides and no default will be set for using when getting
+/// a configuration for a method that has not been given a specific override.
+/// Use ``setDefaultConfiguration(_:forService:)`` to set a specific override for a whole
+/// service, or set a default configuration for all methods by calling ``setDefaultConfiguration(_:)``.
+///
+/// Use the subscript to get and set configurations for specific methods.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public struct MethodConfigurations: Sendable, Hashable {
+  private var elements: [MethodDescriptor: MethodConfiguration]
+
+  /// Create a new ``MethodConfigurations`` with no overrides and no default configuration.
+  public init() {
+    self.elements = [:]
+  }
+
+  /// Get or set the corresponding ``MethodConfiguration`` for the given ``MethodDescriptor``.
+  ///
+  /// Configuration is hierarchical and can be set per-method, per-service
+  /// (``setDefaultConfiguration(_:forService:)``) and globally (``setDefaultConfiguration(_:)``).
+  /// This subscript sets the per-method configuration but retrieves a configuration respecting
+  /// the hierarchy. If no per-method configuration is present, the per-service configuration is
+  /// checked and returned if present. If the per-service configuration isn't present then the
+  /// global configuration is returned, if present.
+  ///
+  /// - Parameters:
+  ///  - descriptor: The ``MethodDescriptor`` for which to get or set a ``MethodConfiguration``.
+  public subscript(_ descriptor: MethodDescriptor) -> MethodConfiguration? {
```

Comment:
Does it make sense to subscript for a service in the future? Then I think making the subscript argument label external makes sense.
```suggestion
  public subscript(descriptor: MethodDescriptor) -> MethodConfiguration? {
```